pkg/reconciler/kubequota/kubequota_controller.go

```diff
@@ -0,0 +1,384 @@
+
+	if err := c.startQuotaForClusterWorkspace(ctx, clusterName); err != nil {
+		return fmt.Errorf("error starting quota controller for cluster %q: %w", clusterName, err)
+	}
+
+	return nil
+}
+
+func (c *Controller) startQuotaForClusterWorkspace(ctx context.Context, clusterName logicalcluster.Name) error {
+	resourceQuotaControllerClient := c.kubeClusterClient.Cluster(clusterName)
+
+	// TODO(ncdc): find a way to support the default configuration. For now, don't use it, because it is difficult
+	// to get support for the special evaluators for pods/services/pvcs.
+	// listerFuncForResource := generic.ListerFuncForResourceFunc(scopedInformerFactory.ForResource)
+	// quotaConfiguration := install.NewQuotaConfigurationForControllers(listerFuncForResource)
+	quotaConfiguration := generic.NewConfiguration(nil, install.DefaultIgnoredResources())
+
+	resourceQuotaControllerOptions := &resourcequota.ControllerOptions{
+		QuotaClient: resourceQuotaControllerClient.CoreV1(),
+		ResourceQuotaInformer: &SingleClusterResourceQuotaInformer{
+			clusterName:     clusterName,
+			delegate:        c.resourceQuotaInformer,
+			registerHandler: c.registerResourceQuotaEventHandlerLocked,
+		},
+		ResyncPeriod:    controller.StaticResyncPeriodFunc(c.quotaRecalculationPeriod),
+		InformerFactory: c.scopingGenericSharedInformerFactory.ForCluster(clusterName),
+		ReplenishmentResyncPeriod: func() time.Duration {
+			return c.fullResyncPeriod
+		},
+		DiscoveryFunc:        c.dynamicDiscoverySharedInformerFactory.DiscoveryData,
+		IgnoredResourcesFunc: quotaConfiguration.IgnoredResources,
+		InformersStarted:     c.informersStarted,
+		Registry:             generic.NewRegistry(quotaConfiguration.Evaluators()),
+		ClusterName:          clusterName,
+	}
+	if resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		if err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage(clusterName.String()+"-resource_quota_controller", resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return err
+		}
+	}
+
+	resourceQuotaController, err := resourcequota.NewController(resourceQuotaControllerOptions)
+	if err != nil {
+		return err
+	}
+
+	// Here we diverge from what upstream does. Upstream starts a goroutine that retrieves discovery every 30 seconds,
+	// starting/stopping dynamic informers as needed based on the updated discovery data. We know that kcp contains
+	// the combination of built-in types plus CRDs. We use that information to drive what quota evaluates.
+
+	// Use a buffered channel so the discoveryChangeNotifier (ddsif) can always enqueue at least 1 notification.
+	discoveryChangedCh := make(chan struct{}, 1)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-discoveryChangedCh:
+				klog.V(4).Infof("%s: got API change notification", clusterName)
+				resourceQuotaController.KCPUpdateMonitors(ctx, clusterName, c.dynamicDiscoverySharedInformerFactory.DiscoveryData)
```

Comment:
Yeah esp since this is going to wait for the informers to sync in the critical path between consuming discovery events we should ensure the send-side of discoveryChangedCh can handle timeouts or something

Reply:
My thinking is:
- Subscriber should use a buffered channel (as in the previous comment, happy to change this to return a new buffered channel to a subscriber)
- Sender (ddsif) uses the select pattern to try to send to the channel, with a default clause to avoid any blocks (and therefore dropping the notification)
- Because the channel is buffered, when the ddsif wants to send a notification, one of the following two statements must be true:
    1. The channel has capacity to receive, so the send to the channel will succeed, and then the subscriber in theory will pick it up
    2. The channel is full, so it's safe to drop the send to the channel (the subscriber hasn't picked up the notification yet, so there is no need to notify it again; i.e., deduplicate)

Reply:
See my comments. Ddsif should not care about consumers and how they implement this. Make it a contract that the handler func does not block.

Reply:
Resolving this thread as it is essentially a dupe of the previous one (which I'll keep open until I address the 🛑)